From e503f457f370602d80cba45aa916da9889bc0aac Mon Sep 17 00:00:00 2001
From: Ikey Doherty <michael.i.doherty@intel.com>
Date: Wed, 14 Oct 2015 17:40:56 +0100
Subject: [PATCH 17/44] bootctl: Handle gummiboot -> systemd migration

On Linux systems we may encounter cases where the legacy gummiboot binary
is still present on the ESP. In these cases a `bootctl install` or even
a `bootctl upgrade` will fail on files such as BOOTX64.EFI, as their internal
version string will not match that of systemd.

To counteract this upgrade problem, we will always continue to replace the
legacy gummiboot (<=48) with systemd's merged tree, to provide a valid upgrade
route for distributions.
---
 src/boot/bootctl.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/boot/bootctl.c b/src/boot/bootctl.c
index 87b1a73c55..178c06d45b 100644
--- a/src/boot/bootctl.c
+++ b/src/boot/bootctl.c
@@ -446,10 +446,13 @@ static int version_check(int fd_from, const char *from, int fd_to, const char *t
         r = get_file_version(fd_to, &b);
         if (r < 0)
                 return r;
-        if (r == 0 || compare_product(a, b) != 0)
-                return log_notice_errno(SYNTHETIC_ERRNO(EEXIST),
-                                        "Skipping \"%s\", since it's owned by another boot loader.",
-                                        to);
+
+        if (r == 0 || compare_product(a, b) != 0) {
+                if (!compare_product(b, "gummiboot"))
+                        return log_notice_errno(SYNTHETIC_ERRNO(EEXIST),
+                                                "Skipping \"%s\", since it's owned by another boot loader.",
+                                                to);
+        }
 
         if (compare_version(a, b) < 0)
                 return log_warning_errno(SYNTHETIC_ERRNO(ESTALE), "Skipping \"%s\", since a newer boot loader version exists already.", to);
-- 
2.21.0

